feat(inlay-hints): add inlay hints global state and toggle function

TheFedaikin opened the pull request:
Closes #345

TheFedaikin commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -4,7 +4,7 @@ local M = {}
 
 function M.new()
   M.namespace = vim.api.nvim_create_namespace("textDocument/inlayHints")
-  local self = setmetatable({ cache = {}, enabled = false }, { __index = M })
```
I am not that versed in internals, so I have no idea why and how this state was used. All I know that I wasn't able to read it externally.

TheFedaikin commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -34,11 +35,22 @@ end
 -- Enable auto hints and set hints for the current buffer
 function M.enable(self)
-  self.enabled = true
+  self.enable = false
```
Is this needed at all?

TheFedaikin replied:
It's a function reset when you already have it enabled, so I guess it's good

katlyn-thomas commented:
ooooh, this is exactly what I've been looking for! I'm not the maintainer, but I would add that updating [this section of the readme](https://github.com/simrat39/rust-tools.nvim/blob/master/README.md?plain=1#L73-L87) would be useful.

other than that, I think this is great. I'm already using your branch in my config.

TheFedaikin commented:
I am glad this works for you! Feel free to report if anything breaks, I'll try to patch this PR to fix that.

About the documentation, you are absolutely correct - I think I was too focused on the actual functionality that I forgot to improve that. I'll do that tomorrow, but from the looks of it, this repo is pretty slow nowadays, so I am not expecting this to be merged anytime soon, which is unfortunate.

barthap commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -34,11 +35,22 @@ end
+-- Toggles inlay hints state globally. Uses disable and enable internally
+function M.toggle(self)
+  if self.is_enabled then
+    M.disabled(self)
```
```suggestion
    M.disable(self)
```
Isn't this a typo?

TheFedaikin replied:
Oh wow, I didn't double check the hunk, and committed working, but unclean version. 

Good catch, I updated everything in the PR today, so feel free to report any regressions/problems regarding to new implementation, here

TheFedaikin commented:
Thanks everyone in the interest for this PR!

I updated implementation, created an external command for toggle and addressed the documentation comment. 

Feel free to report regressions or/and add your suggestions for this here, it's been a very helpful feedback so far!

TheFedaikin commented:
I changed the PR name so it's clear what it's current implementation is doing. Originally I was thinking about also exposing `enabled` and `disabled` state externally, but opted out of it. If you can think about the use-case that would need that feature, please let me know, but this works well for me at this moment.

utkarshgupta137 commented:
I'm not sure if @simrat39 is active these days. I've been using https://github.com/lvimuser/lsp-inlayhints.nvim, which already has this option. It works perfectly with rust-tools. It also shows more information I think.

lcrownover commented:
Just wanted to chime in that this feature would be super useful as I'd like to be able to start with them disabled but have the ability to easily show them 😄

MunifTanjim commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -4,7 +4,7 @@ local M = {}
 function M.new()
   M.namespace = vim.api.nvim_create_namespace("textDocument/inlayHints")
-  local self = setmetatable({ cache = {}, enabled = false }, { __index = M })
+  local self = setmetatable({ cache = {} }, { __index = M })
```
```suggestion
  local self = setmetatable({ cache = {}, enabled = false }, { __index = M })
```

TheFedaikin replied:
Does it work because any lookup to the table entry will be falsey by default, therefore we don't need to do defaulting to false explicitly? This is my first Lua PR which is why this looks interesting. These suggestions seem to implicate that we can add these values dynamically in the self table and it will work just fine?

MunifTanjim replied:
Yep, it will work. You can check this commit to try it out: https://github.com/MunifTanjim/rust-tools.nvim/commit/2ba50c9baee2b9ca4456e8c3dd59580182337d5c

But yeah... it's a bit confusing to be honest. There's room for improvements (re-organizing some stuffs to make it easier to understand).

TheFedaikin replied:
I see. This makes sense as most minimal working solution, so I am gonna commit that.

MunifTanjim commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -17,6 +17,7 @@ end
 -- Disable hints and clear all cached buffers
 function M.disable(self)
   self.disable = false
+  M.is_enabled = false
```
```suggestion
  self.enabled = false
```

MunifTanjim commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -34,11 +35,21 @@ end
 -- Enable auto hints and set hints for the current buffer
 function M.enable(self)
-  self.enabled = true
+  self.enable = false
+  M.is_enabled = true
```
```suggestion
  self.enabled = true
```

MunifTanjim commented on lua/rust-tools/inlay_hints.lua:
```diff
@@ -34,11 +35,21 @@ end
 
+-- Toggles inlay hints state globally. Uses disable and enable internally
+function M.toggle(self)
+  if self.is_enabled then
```
```suggestion
  if self.enabled then
```

MunifTanjim reviewed (commented):
No need to add another variable `M.is_enabled`. Just need to fix the existing one. I've added the suggestions.

MunifTanjim commented:
I've pulled this in https://github.com/MunifTanjim/rust-tools.nvim/tree/patched

penso commented:
Damn would love this to be merged.

nico-mayora commented:
How close is this to getting merged?

TheFedaikin commented:
I don't want to be alarmist but I don't think the author have a time for this project and there's absolutely no signs he will be back so it's probably easier to fork and cherry-pick changes that you want for yourself and use it. Alternatively we can wait (or find if it exists) for `official` fork so it can move forward.
